File: javascript/binary-search/binary-search.js
export const find = (array, n) => {
    if (array.length == 0)
        throw "Value not in array";

    if (array.length == 1 && array[0] != n)
        throw "Value not in array";

    let middle = parseInt(array.length / 2);

    if (array[middle] < n)
        return middle + find(array.slice(middle + 1), n) + 1;
    else if (array[middle] > n)
        return find(array.slice(0, middle), n)
    else
        return middle;
};

File: csharp/binary-search/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        if (input.Length == 0)
            return -1;

        if (input.Length == 1) {
            if (input[0] == value)
                return 0;
            else
                return -1;
        }

        var middle = input.Length / 2;

        if (input[middle] < value) {
            var found = Find(input[middle..], value);

            if (found > -1)
                found += middle;

            return found;
        } else if (input[middle] > value) {
            return Find(input[..middle], value);
        } else {
            return middle;
        }
    }
}
